### ISSUE TITLE
Get url authentication from google

### ISSUE THREAD

**Issue by SimonLab** [FAC-GM/gm-magic#21]
As a user I want to be able to click on a "login with google" button which is display on the popup page

**Comment by SimonLab** [comment]
We need to generate a get request from the extension to get the url login. When do we trigger this request, on a click of the button of the extension or when the popup page is loaded

**Comment by SimonLab** [comment]
To trigger an action when the popup page is loaded we can simply executing a js script from the html popup page

**Comment by heron2014** [comment]
With this [commit](https://github.com/FAC-GM/gm-magic/commit/a417049673341bba7e1e694de02a9076b51e21ea) I created button with the google url, but the problem is that when you click on the link it doesn't redirect you anywhere. 

I found possible solution [here](http://stackoverflow.com/questions/24434470/redirect-to-url-using-google-chrome-extension). We are still looking for the right solution.

**Comment by SimonLab** [comment]
Another solution might be to use [identity](https://developer.chrome.com/apps/identity)

**Comment by SimonLab** [comment]
"Web authentication protocols utilize HTTP features, but Chrome Apps run inside the app container; they don’t load over HTTP and can’t perform redirects or set cookies."

**Comment by heron2014** [comment]
Look like we need some time to read documentation: We need extra time : 1h to go through examples.

**Comment by nelsonic** [comment]
@SimonLab do we *need* the Extension to Set a Cookie? 
or, can we store the JWT once Auth is done and send it as a Header when `POST`ing data to the API?

**Comment by SimonLab** [comment]
I don't think we need cookie but I'm also a bit worry about the redirection. This means that we can't even display the google authorize page. That's why we are now looking at the [app identity](https://developer.chrome.com/apps/app_identity) which seems to be the other way to authenticate with google from an app or a chrome extension.

**Comment by SimonLab** [comment]
some links to read: http://stackoverflow.com/questions/25044936/chrome-identity-user-authentication-in-a-chrome-extension

**Comment by SimonLab** [comment]
@nelsonic here a small recap of our thoughts:

- 1) Authenticate the user in the chrome extension. For this I don't think we can reuse the actual authentication we have in our app. I think we can use the special [user authentication](https://developer.chrome.com/apps/app_identity) for app/extension. For this wee need to already publish the extension to obtain a google id (maybe we can reuse an existant key?).

- 2) Our api endpoint will use the jwt scheme to manage authentication. So we need to generate a jwt from the information of the user logged in with the extension. We are still thinking on how to do this to keep the app secure.

- 3) Once we have a jwt we need to find a way to store it ([chrome cookies](https://developer.chrome.com/extensions/cookies)?) and for each request to the api send this token to   our api, our validate function in our app will do the rest.

I'm not sure if I'm clear enough, let us know what you think about these steps.

**Comment by SimonLab** [comment]
more reading: http://zacstewart.com/2015/05/14/using-json-web-tokens-to-authenticate-javascript-front-ends-on-rails.html

**Comment by SimonLab** [comment]
One possible solution is as soon as we are authenicate on the chrome extension with the identity chrome api, we could try to generate the right JWT, for that we need:
- a front end jwt sign function to encode the jwt, it implies to have a "heavier" chrome extension as we are adding a package to encrypt JWT.
- a secret key on the extension which **must** be keep secret! Is it possible to store in a chrome extension a secret value, like an private environment variable for the extension?

**Comment by heron2014** [comment]
Proposition of the solution : authenticate via chrome.identity api , send token in a header to server and on a server generate JWT using received token from the header.

**Comment by SimonLab** [comment]
@heron2014 This solution is not secure as someone can try to guess the token you are sending to the server and get a valid JWT

**Comment by nelsonic** [comment]
@SimonLab + @heron2014 that sounds like it will work. have you started implementing?

**Comment by SimonLab** [comment]
@nelsonic we are starting to test the authentication directly with the chrome api. But I'm still unsure on how to generate a valid JWT, on the front end (need to store safely the jwt secret, is it possible?) on the server (we can't create an endpoint it's not secure)?

**Comment by nelsonic** [comment]
It might be best if you do a Google Docs Drawing for this so we can *all* understand the thought process.

**Comment by SimonLab** [comment]
:+1:

**Comment by iteles** [comment]
@SimonLab @heron2014 As we discussed, it would be great if you could update this issue with the latest (e.g. URLs to the drawing) as part of your end of day wrap-up.

**Comment by SimonLab** [comment]
![stateextension](https://cloud.githubusercontent.com/assets/6057298/11310879/335722c2-8fc4-11e5-955a-c05dc65897fa.png)

This is our last thougths on how to get a JWT for the extension. We need to do more exploration on how to generate the JWT, maybe we will find a way to do the fist try.

**Comment by heron2014** [comment]
I found on stack overflow possible solution to our issue: http://stackoverflow.com/questions/7947038/how-to-authenticate-with-google-via-oauth-2-0-in-a-popup

**Comment by SimonLab** [comment]
We are currently trying to use the authentication with the chrome identity api. We still have some research and documentation on the problem and we don't have a direct solution yet
